python/gstaichi/lang/ast/ast_transformer_utils.py

```diff
@@ -195,6 +195,10 @@ def __init__(
         is_real_function: bool,
         autodiff_mode: AutodiffMode,
         raise_on_templated_floats: bool,
+        # during 1st pass, we collect the names of used parameters
+        used_py_dataclass_parameters_collecting: set[str],
+        # during 2nd pass, we only handle these names
+        used_py_dataclass_parameters_enforcing: set[str] | None,
```

Comment:
Could you elaborate why distinguishing parameters during collection and enforcing phases is never ? The same buffer cannot be used for both? since I have the feeling these 2 phases come one after the other and never simultaneously?

Reply:
(basically, there were tons of bugs where I was mutating the enforcing buffer by accident, or using the collecting buffer for enforcement etc, when I had only a single buffer; was just super confusing I felt).